pretty_midi/instrument.py

```diff
@@ -205,6 +205,93 @@ def get_end_time(self):
+                                  normalize=False):
+        """Computes the frequency of pitch classes of the current instrument,
+        optionally weighted by their durations or velocities.
+
+        Parameters
+        ----------
+        use_duration : bool
+            Weight frequency by note duration
+        use_velocity : bool
+            Weight frequency by note velocity
+        normalize : bool
+            Normalizes the histogram such that the sum of bin values is 1.
+
+        Returns
+        -------
+        histogram : np.ndarray, shape=(12,)
+            Histogram of pitch classes given current instrument, optionally
+            weighted by their durations or velocities
+        """
+
+        # Return all zeros if track is drum
+        if self.is_drum:
+            return np.zeros(12)
+
+        weights = np.ones(len(self.notes))
+
+        # Assumes that duration and velocity have equal weight
+        if use_duration:
+            weights *= [note.end - note.start for note in self.notes]
+        if use_velocity:
+            weights *= [note.velocity for note in self.notes]
+
+        histogram, _ = np.histogram([n.pitch % 12 for n in self.notes],
+                                    bins=np.arange(13),
+                                    weights=weights,
+                                    density=normalize)
+
+        return histogram
+
+    def get_pitch_class_transition_matrix(self, normalize=False):
+        """Computes the pitch class transition matrix of the current instrument
+
+        Transitions are added whenever the end of a note is within 50ms from the
+        start of any other note.
+
+        Parameters
+        ----------
+        normalize : bool
+            Normalize transition matrix such that matrix sum equals is 1.
+
+        Returns
+        -------
+        pitch_class_transition_matrix : np.ndarray, shape=(12,12)
+            Pitch class transition matrix
+        """
+
+        # instrument is drum or less than one note, return all zeros
+        if self.is_drum or len(self.notes) <= 1:
+            return np.zeros((12, 12))
+
+        # container for holding pitch class transitions
+        pitch_class_transition_matrix = np.zeros((12, 12))
+
+        # use 20hz(0.05s) as the minimun time threshold for separate notes
+        time_thresh = 0.05
+
+        # data matrix
+        data_mat = np.array([[x.start, x.end, x.pitch % 12]
+                             for x in self.notes])
+
+        # compute distance matrix for all end and start time pairs
+        dist_mat = np.subtract.outer(data_mat[:, 1], data_mat[:, 0])
+
+        # compute boolean matrix of indices within threshold
+        trans_idx= abs(dist_mat) < time_thresh
+
+        # update transition matrix
+        for i in xrange(len(data_mat)):
+            pitch_class_transition_matrix[
+                data_mat[i, 2], data_mat[trans_idx[i], 2].astype(int)] += 1
```

Comment:
I like the idea here, but I don't think this will do what you want it to do; specifically, if you end up constructing multiple index tuples which are the same in one of these loop iterations, only 1 will get added to it.  For example,
```Python
>>> a = np.zeros((4, 4))
>>> a[[2, 2, 2, 2], [3, 3, 3, 3]] += 1
>>> print a

[[ 0.  0.  0.  0.]
 [ 0.  0.  0.  0.]
 [ 0.  0.  0.  1.]
 [ 0.  0.  0.  0.]]
```
I think what basically means that if note `i` transitions to the same pitch multiple times throughout the instrument, the pitch class transition matrix will only get one transition added.  You can confirm this by noting that `trans_idx.sum() != pitch_class_transition_matrix.sum()` (i.e. the number of notes whose start times are within the tolerance of another notes end time is not the same as the number of transitions recorded).

One way to do all this, which I think will be hard to beat in terms of efficiency, is with `np.histogram2d`:

```Python
trans_idx = np.argwhere(abs(dist_mat) < time_thresh)
pitch_class_transition_matrix = np.histogram2d(
    data_mat[trans_idx[:, 0], 2],
    data_mat[trans_idx[:, 1], 2], bins=12)[0]
```

Semantically, though, I like this the most:

```Python
# instrument is drum or less than one note, return all zeros
if self.is_drum or len(self.notes) <= 1:
    return np.zeros((12, 12))

# use 20hz(0.05s) as the minimun time threshold for separate notes
time_thresh = 0.05

# Retrieve note starts, ends, and transitions from self.notes
starts, ends, transitions = np.array(
    [[x.start, x.end, x.pitch % 12] for x in self.notes]).T

# Find indices of notes pairs of notes where the end time of
# one note is within time_thresh of the start time of the other
sources, targets = np.argwhere(abs(dist_mat) < time_thresh).T

return np.histogram2d(
    transitions[sources], transitions[targets], bins=12)[0]
```